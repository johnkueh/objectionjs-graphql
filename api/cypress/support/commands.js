Cypress.Commands.add('deleteUser', email => {
  return cy.task('deleteUser', email);
});

Cypress.Commands.add('factory', args => {
  return cy.task('factory', args);
});

Cypress.Commands.add('createUserAndLogin', args => {
  cy.task('factory', {
    method: 'create',
    type: 'user',
    args,
    include: ['jwt']
  }).then(user => {
    cy.setCookie('jwt', user.jwt);
    cy.wrap(user);
  });
});

Cypress.Commands.add('createUserWithWorkspacesAndLogin', args => {
  cy.task('factory', {
    method: 'create',
    type: 'userWithWorkspace',
    args,
    include: ['jwt', 'workspaces']
  }).then(user => {
    cy.setCookie('jwt', user.jwt);
    cy.wrap(user);
  });
});
