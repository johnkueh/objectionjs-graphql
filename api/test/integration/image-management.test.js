import '../support/transactional-tests';
import '../factories';

import factory from 'factory-girl';
import request from '../support/request';
import handler, { path } from '../../src/index';

describe('Upserting images', () => {
  let user;
  const query = `
    mutation($input: UpsertImageInput!) {
      upsertImage(input: $input) {
        id
        publicId
        caption
      }
    }
  `;

  beforeEach(async () => {
    user = await factory.create('user', {
      email: 'john@doe.com',
      password: 'password'
    });
  });

  it('is not able to upsert image without auth', async () => {
    const res = await request({
      handler,
      apiPath: path,
      query,
      variables: {
        input: {
          publicId: 'public-id',
          caption: 'A test caption'
        }
      }
    });

    expect(res.errors[0].extensions.exception.errors).toEqual({
      auth: 'You are not authorized to perform this action'
    });
  });

  it('is not able to create an image with missing fields', async () => {
    const res = await request({
      handler,
      apiPath: path,
      cookies: [`jwt=${user.jwt}`],
      query,
      variables: {
        input: {
          caption: 'A test caption'
        }
      }
    });

    expect(res.errors[0].extensions.exception.errors).toEqual({
      publicId: 'PublicId is a required field',
      imageableType: 'ImageableType is a required field',
      imageableId: 'ImageableId is a required field'
    });
  });

  it('is not able to create an image with invalid fields', async () => {
    const res = await request({
      handler,
      apiPath: path,
      cookies: [`jwt=${user.jwt}`],
      query,
      variables: {
        input: {
          publicId: 'xxx-id',
          imageableId: 'yyy-id',
          imageableType: 'WeirdImage',
          caption: 'A test caption'
        }
      }
    });

    expect(res.errors[0].extensions.exception.errors.imageableType).toContain(
      'ImageableType must match the following'
    );
  });

  it('is not able to create an image with existing publicId', async () => {
    await factory.create('image', {
      publicId: 'existing-public-id',
      imageableType: 'UserLogo',
      imageableId: user.id
    });

    const res = await request({
      handler,
      apiPath: path,
      cookies: [`jwt=${user.jwt}`],
      query,
      variables: {
        input: {
          publicId: 'existing-public-id',
          imageableId: 'yyy-id',
          imageableType: 'UserLogo'
        }
      }
    });

    expect(res.errors[0].extensions.exception.errors).toEqual({
      publicId: 'PublicId is already taken'
    });
  });

  it('is able to create an image with valid fields', async () => {
    const res = await request({
      handler,
      apiPath: path,
      cookies: [`jwt=${user.jwt}`],
      query,
      variables: {
        input: {
          publicId: 'xxxpublicid',
          caption: 'A test caption',
          imageableType: 'UserLogo',
          imageableId: user.id
        }
      }
    });

    expect(res.data.upsertImage).toEqual({
      id: expect.any(String),
      publicId: 'xxxpublicid',
      caption: 'A test caption'
    });

    const userLogo = await user.$relatedQuery('logo');
    expect(userLogo.publicId).toEqual('xxxpublicid');
    expect(userLogo.caption).toEqual('A test caption');
  });

  it('is able to update an existing image with valid fields, does not create a new image', async () => {
    const image = await factory.create('image', {
      publicId: 'existing-public-id',
      imageableType: 'UserLogo',
      imageableId: user.id
    });

    const res = await request({
      handler,
      apiPath: path,
      cookies: [`jwt=${user.jwt}`],
      query,
      variables: {
        input: {
          id: image.id,
          publicId: 'new-public-id',
          imageableType: 'UserLogo',
          imageableId: user.id
        }
      }
    });

    expect(res.data.upsertImage).toEqual({
      id: image.id,
      caption: null,
      publicId: 'new-public-id'
    });

    const userLogo = await user.$relatedQuery('logo');
    expect(userLogo.id).toEqual(image.id);
    expect(userLogo.publicId).toEqual('new-public-id');
  });
});

describe('Deleting images', () => {
  const query = `
    mutation($input: DeleteImageInput!) {
      deleteImage(input: $input) {
        count
      }
    }
  `;

  beforeEach(async () => {
    await factory.create('user', {
      email: 'john@doe.com',
      password: 'password'
    });
  });

  it('is not able to delete image without auth', async () => {
    const res = await request({
      handler,
      apiPath: path,
      query,
      variables: {
        input: {
          id: 'xxx-image-id'
        }
      }
    });

    expect(res.errors[0].extensions.exception.errors).toEqual({
      auth: 'You are not authorized to perform this action'
    });
  });
});
