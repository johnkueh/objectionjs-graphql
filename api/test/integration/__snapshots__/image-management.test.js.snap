// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deleting images is not able to delete image without auth 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "auth": "You are not authorized to perform this action",
    },
  },
}
`;

exports[`Upserting images is able to create an image with existing imageableId but different scope 1`] = `
Object {
  "data": Object {
    "upsertImage": Object {
      "caption": null,
      "id": Any<String>,
      "publicId": "xxxpublicid",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Upserting images is able to create an image with valid fields 1`] = `
Object {
  "data": Object {
    "upsertImage": Object {
      "caption": "A test caption",
      "id": Any<String>,
      "publicId": "xxxpublicid",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Upserting images is able to update an existing image with valid fields, does not create a new image 1`] = `
Object {
  "data": Object {
    "upsertImage": Object {
      "caption": null,
      "id": Any<String>,
      "publicId": "new-public-id",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Upserting images is not able to create an image with existing imageableId 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "imageableId": "ImageableId with scope imageableType is already taken",
    },
  },
}
`;

exports[`Upserting images is not able to create an image with existing publicId 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "publicId": "PublicId is already taken",
    },
  },
}
`;

exports[`Upserting images is not able to create an image with invalid fields 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "imageableType": "ImageableType must match the following: \\"/(UserLogo|UserProfilePhoto)/\\"",
    },
  },
}
`;

exports[`Upserting images is not able to create an image with missing fields 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "imageableId": "ImageableId is a required field",
      "imageableType": "ImageableType is a required field",
      "publicId": "PublicId is a required field",
    },
  },
}
`;

exports[`Upserting images is not able to upsert image without auth 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "auth": "You are not authorized to perform this action",
    },
  },
}
`;
